behind-the-veil-siteroot/imports/ui/components/pages/messages/MessagesPage.jsx

```diff
@@ -1,50 +1,220 @@
+    const isLoadingUserImages = useSubscribe('profile_images');
+    const isLoading = isLoadingChats() || isLoadingUserImages();
+    
+    // get data from db
+    let chatsData = useTracker(() => {
+        return ChatCollection.find({
+            $or: [
+                { brideUsername: userInfo.username },
+                { artistUsername: userInfo.username }
+            ]
+        }).fetch();
+    });
+
+    let usersImagesData = useTracker(() => {
+        return ImageCollection.find().fetch;
+    })
+    
+    // manual aggregation into chatsData with the other user's images
+    for (let i = 0; i < chatsData.length; i++) {
+        // find the other user's username
+        const otherUser = getOtherUsername(userInfo.username, chatsData[i]);
+
+        for (let j = 0; j < usersImagesData.length; j++) {
+            // find the other user's image and add their image to the chat data
+            if (usersImagesData[j].targetId == otherUser) {
+                chatsData[i].otherUserImage = usersImagesData[j].imageData;
+                break;
+            }
+        }
+    }
+
+    // sort chats data in descending order for the dates
+    chatsData.sort(function(dateOne, dateTwo) {
+        let keyOne = dateOne.chatUpdatedDate;
+        let keyTwo = dateTwo.chatUpdatedDate;
+        // Compare the 2 dates
+        if (keyOne > keyTwo) return -1;
+        if (keyOne < keyTwo) return 1;
+        return 0;
+    });
+
+    // set the default conversation to the conversation that was updated most recently
     const [selectedConversationIndex, setSelectedConversationIndex] = useState(0);  // track the conversation selected, initially 0
-    const handlePreviewClick = (index) => {
+    
+    // if the user is already in a conversation with someone (based on the url), show the conversation with 
+    // that user. if it doesn't exist, create a new chat with that user.
+    useEffect(() => {
+        const otherUsername = window.location.hash.slice(1);
+
+        // helper function to find an existing chat
+        const findChatIndex = (username) => {
+            return chatsData.findIndex(chat => 
+                username === chat.artistUsername || username === chat.brideUsername
+            );
+        };
+
+        if (isLoadingChats()) return; // return if chats are still loading
+
+        // for the default message url, route to default chat
+        if (otherUsername === "") {
+            routeToDefaultChat();
+            return;
+        }
+
+        const chatIndex = findChatIndex(otherUsername);
+        // if chat is found, set it as the selected conversation
+        if (chatIndex !== -1) {
+            setSelectedConversationIndex(chatIndex);
+            // set the read status of the conversation to true
+            chatRead(chatIndex, true);
+            
+        } else {
+            createNewChat(otherUsername);
+        }
+    }, [chatsData, window.location.hash.slice(1), isLoadingChats()]);
```

Comment:
Just curious bc I haven't come across this before, but is 'isLoadingChats' supposed to be called in this dependency array, or was the intention to include the 'isLoadingChats' function in the dependency array instead of the returned value?

Reply:
The intention was to have the returned value so that the useEffect stuff gets called when the loading status of the chats changes but I now realise it might not be functioning as it's supposed to, I should probably replace it with isLoading (the variable) - would this be the right approach?

Reply:
Yeah i think that's the way to go. 

I think also, if you only want to listen for the chats loading state and not the user images too, then you can store the state that isLoadingChats returns at the top of the component, and then put that in the dependency array like:

```
const isLoadingChats = useSubscribe('all_user_chats', userInfo.username);
const isLoadingChatsState = isLoadingChats()
const isLoading = isLoadingChatsState || isLoadingUserImages();
...


useEffect( () => {
     if (isLoadingChatsState) return;
    // rest of logic here
    
}, [chatsData, window.location.hash.slice(1), isLoadingChatsState]);
```

take w a grain of salt though, i haven't tested this